src/psyclone/psyir/backend/sir.py

```diff
@@ -0,0 +1,408 @@
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford, STFC Daresbury Lab.
+
+'''SIR PSyIR backend. Generates SIR code from PSyIR nodes. Currently
+limited to PSyIR Kernel schedules as PSy-layer PSyIR already has a
+gen() method to generate Fortran.
+
+'''
+
+from psyclone.psyir.backend.base import PSyIRVisitor, VisitorError
+from psyclone.psyGen import Reference, BinaryOperation, Literal, \
+    Array, UnaryOperation
+from psyclone.nemo import NemoLoop, NemoKern
+
+
+def gen_stencil(node):
+    '''Given an array access as input, determine the form of stencil
+    access and return it in the form expected by the SIR as a
+    string. Raise an exception if the array access is not a recognised
+    stencil access.
+
+    :param node: an array access.
+    :type node: :py:class:`psyclone.psyGen.Array`
+
+    :returns: the SIR stencil access format for the array access.
+    :rtype: str
+
+    :raises VisitorError: if the node is not the expected type or the \
+    array access is not in a recognised stencil form.
+
+    '''
+    if not isinstance(node, Array):
+        raise VisitorError(
+            "gen_stencil expected an Array as input but found '{0}'."
+            "".format(type(node)))
+    dims = []
+    for child in node.children:
+        if isinstance(child, Reference):
+            dims.append("0")
+        elif isinstance(child, BinaryOperation):
+            if isinstance(child.children[0], Reference) and \
+               isinstance(child.children[1], Literal):
+                if child.operator.name == "SUB":
```

Comment:
Good point, I wrote this before Sergi's changes ( I hope :-) ).

1+i is not a supported (recognised) stencil access, it has to conform to 'variable +/- value', otherwise an exception will be raised, see 'test_gen_stencil_3' in 'sir_test.py'.